e2e/helm_client.go

```diff
@@ -0,0 +1,286 @@
+/*
+Copyright 2017 The Kubernetes Authors All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// HelmManager provides functionality to install client/server helm and use it
+type HelmManager interface {
+	// InstallTiller will bootstrap tiller pod in k8s
+	InstallTiller() error
+	// DeleteTiller removes tiller pod from k8s
+	DeleteTiller(removeHelmHome bool) error
+	// Install chart, returns releaseName and error
+	Install(chartName string, values map[string]string) (string, error)
+	// Status verifies state of installed release
+	Status(releaseName string) error
+	// Delete release
+	Delete(releaseName string) error
+	// Upgrade release
+	Upgrade(chartName, releaseName string, values map[string]string) error
+	// Rollback release
+	Rollback(releaseName string, revision int) error
+}
+
+// BinaryHelmManager uses helm binary to work with helm server
+type BinaryHelmManager struct {
+	Clientset         kubernetes.Interface
+	Namespace         string
+	HelmBin           string
+	TillerHost        string
+	UseCanary         bool
+	UseServiceAccount bool
+}
+
+func (m *BinaryHelmManager) InstallTiller() error {
+	arg := make([]string, 0, 5)
+	var err error
+	arg = append(arg, "init", "--tiller-namespace", m.Namespace)
+	if m.UseCanary {
+		arg = append(arg, "--canary-image")
+	}
+	if m.UseServiceAccount {
+		arg = append(arg, "--service-account", "tiller")
+		if err = m.InstallServiceAccounts(); err != nil {
+			return err
+		}
+	}
+	_, err = m.executeUsingHelm(arg...)
+	if err != nil {
+		return err
+	}
+	By("Waiting for tiller pod")
+	waitTillerPod(m.Clientset, m.Namespace)
+	return nil
+}
+
+func (m *BinaryHelmManager) DeleteTiller(removeHelmHome bool) error {
+	arg := make([]string, 0, 4)
```

Comment:
Nit: If you are appending more than the capacity you are setting here, why not just do `arg := []string{}`?

Reply:
True, will fix